src/pip/_internal/commands/install.py

```diff
@@ -546,63 +536,83 @@ def site_packages_writable(**kwargs):
+    """
+    # Check for incompatible options
+    locations = {"--target": target_dir, "--user": use_user_site,
+                 "--root": root_path, "--prefix": prefix_path}
+    destinations = [k for k, v in locations.items() if v]
+    if len(destinations) > 1:
+        last, rest = destinations[-1], ", ".join(destinations[:-1])
+        raise CommandError("Different installation locations are implied by "
+                           "{} and {}.".format(rest, last))
+
+    if target_dir:
+        if os.path.exists(target_dir) and not os.path.isdir(target_dir):
+            raise InstallationError("Target path exists but is not "
+                                    "a directory: {}".format(target_dir))
+        if not test_writable_dir(target_dir):
             raise InstallationError(
-                "Can not perform a '--user' install. User site-packages "
-                "are not visible in this virtualenv."
-            )
-        logger.debug("User install by explicit request")
-        return True
-
-    # If we are here, user installs have not been explicitly requested/avoided
-    assert use_user_site is None
-
-    # user install incompatible with --prefix/--target
-    if prefix_path or target_dir:
-        logger.debug("Non-user install due to --prefix or --target option")
+                "Cannot write to target directory: {}".format(target_dir))
+        logger.debug("Non-user install due to specified target directory")
         return False
-
-    # If user installs are not enabled, choose a non-user install
-    if not site.ENABLE_USER_SITE:
-        logger.debug("Non-user install because user site-packages disabled")
+    if prefix_path:
+        if not test_writable_dir(prefix_path):
+            raise InstallationError(
+                "Cannot write to prefix path: {}".format(prefix_path))
+        logger.debug("Non-user install due to specified prefix path")
         return False
 
-    # If we have permission for a non-user install, do that,
-    # otherwise do a user install.
-    if site_packages_writable(root=root_path, isolated=isolated_mode):
+    if use_user_site is not None:
+        use_user_site = bool(use_user_site)
+        logger.debug("{} install by explicit request".format(
+            "User" if use_user_site else "Non-user"))
+        if use_user_site is True and site.ENABLE_USER_SITE is None:
+            raise InstallationError("User site-packages are disabled for "
+                                    "security reasons or by an administrator.")
+    elif site.ENABLE_USER_SITE is None:
+        logger.debug("Non-user install since user site-packages are disabled "
+                     "for security reasons or by an administrator.")
+        use_user_site = False
+    elif site.ENABLE_USER_SITE is False:
+        logger.debug("Non-user install since user site-packages are disabled "
+                     "by user request (with 'python -s' or PYTHONNOUSERSITE)")
+        use_user_site = False
```

Comment:
You could unite these blocks into one and reuse the named condition from https://github.com/pypa/pip/pull/7828/files#r389261764:
```suggestion
    elif not site.ENABLE_USER_SITE:
        disabled_reason_msg = (
            "for security reasons or by an administrator"
            if is_user_site_disabled_for_security else
            "by user request (with 'python -s' or PYTHONNOUSERSITE)"
        )
        logger.debug(
            "Non-user install since user site-packages are disabled {reason}.".
            format(reason=disabled_reason_msg),
        )
        use_user_site = False
```

Reply:
I'm not sure about this, it makes the code overly engineered.  Additionally, (also replying to [an other review](https://github.com/pypa/pip/pull/7828/files#r389261764)), `site.ENABLE_USER_SITE` is part of the standard library, aliasing it introduces new variables which is more difficult to track and it is not easy to name them in a way that describe what the flag really means.  IMHO putting an comment briefly reminding of `site.ENABLE_USER_SITE` meaning would better serve clarity.